metric_tank/cluster.go

```diff
@@ -0,0 +1,213 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bitly/go-hostpool"
+	"github.com/grafana/grafana/pkg/log"
+	met "github.com/grafana/grafana/pkg/metric"
+	"github.com/nsqio/go-nsq"
+	"github.com/raintank/raintank-metric/app"
+	"github.com/raintank/raintank-metric/instrumented_nsq"
+)
+
+var (
+	hostPool      hostpool.HostPool
+	producers     map[string]*nsq.Producer
+	clusterStatus *ClusterStatus
+)
+
+type ClusterStatus struct {
+	sync.Mutex
+	Instance   string
+	Primary    bool
+	LastChange time.Time
+}
+
+func NewClusterStatus(instance string, initialState bool) *ClusterStatus {
+	return &ClusterStatus{
+		Instance:   instance,
+		Primary:    initialState,
+		LastChange: time.Now(),
+	}
+}
+
+func (c *ClusterStatus) Marshal() ([]byte, error) {
+	c.Lock()
+	defer c.Unlock()
+	return json.Marshal(c)
+}
+
+func (c *ClusterStatus) Set(newState bool) {
+	c.Lock()
+	defer c.Unlock()
+	c.Primary = newState
+	c.LastChange = time.Now()
+}
+
+func (c *ClusterStatus) IsPrimary() bool {
+	c.Lock()
+	defer c.Unlock()
+	return c.Primary
+}
+
+type PersistMessage struct {
+	Instance string `json:"instance"`
+	Key      string `json:"key"`
+	T0       uint32 `json:"t0"`
+}
+
+func (msg *PersistMessage) Send() {
+	if *topicNotifyPersist == "" {
+		return
+	}
+	body, err := json.Marshal(&msg)
+	if err != nil {
+		log.Fatal(4, "failed to marshal persistMessage to json.")
+	}
+	hostPoolResponse := hostPool.Get()
+	p := producers[hostPoolResponse.Host()]
+	err = p.Publish(*topicNotifyPersist, body)
+	if err != nil {
+		log.Warn("publisher marking host %s as faulty due to %s", hostPoolResponse.Host(), err)
+		hostPoolResponse.Mark(err)
+		msg.Send()
+	}
+	return
+}
+
+type MetricPersistHandler struct {
+	metrics Metrics
+}
+
+func NewMetricPersistHandler(metrics Metrics) *MetricPersistHandler {
+	return &MetricPersistHandler{
+		metrics: metrics,
+	}
+}
+
+func (h *MetricPersistHandler) HandleMessage(m *nsq.Message) error {
+	ms := PersistMessage{}
+	err := json.Unmarshal(m.Body, &ms)
+	if err != nil {
+		log.Error(3, "skipping message. %s", err)
+		return nil
+	}
+	if ms.Instance == *instance {
+		log.Debug("skipping message we generated. %s - %s:%d", ms.Instance, ms.Key, ms.T0)
+		return nil
+	}
+
+	// get metric
+	if agg, ok := metrics.Get(ms.Key); ok {
```

Comment:
yes, this is a known problem.  Though not something that can be easily fixed.  When we move to Kafka this should be much easier to address as we can include the partition offset that the primary node is up to, allowing the secondary nodes to determine if they are up to date in processing the metric data before attempting to sync the save state.